ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctKudaf.java

```diff
@@ -45,90 +45,54 @@
     super(
-        functionName,
-        argIndexInValue,
-            () -> (T[]) Array.newInstance(ttClass, tkVal),
+        functionName, argIndexInValue,
+        ArrayList::new,
         SchemaBuilder.array(outputSchema).build(),
         Collections.singletonList(outputSchema)
     );
 
     this.tkVal = tkVal;
     this.ttClass = ttClass;
     this.outputSchema = outputSchema;
-    this.comparator = (v1, v2) -> {
-      if (v1 == null && v2 == null) {
-        return 0;
-      }
-      if (v1 == null) {
-        return 1;
-      }
-      if (v2 == null) {
-        return -1;
-      }
-
-      return Comparator.<T>reverseOrder().compare(v1, v2);
-    };
   }
 
   @Override
-  public T[] aggregate(final T currentValue, final T[] aggregateValue) {
-    if (currentValue == null) {
-      return aggregateValue;
-    }
-
-    final T last = aggregateValue[aggregateValue.length - 1];
-    if (last != null && currentValue.compareTo(last) <= 0) {
-      return aggregateValue;
+  public List<T> aggregate(final T currentVal, final List<T> currentAggValList) {
+    if (currentVal == null) {
+      return currentAggValList;
     }
-
-    if (ArrayUtil.containsValue(currentValue, aggregateValue)) {
-      return aggregateValue;
+    final int currentSize = currentAggValList.size();
+    if (!currentAggValList.isEmpty()) {
+      final T last = currentAggValList.get(currentSize - 1);
+      if (currentVal.compareTo(last) <= 0
+          && currentSize == tkVal) {
+        return currentAggValList;
+      }
     }
-
-    final int nullIndex = ArrayUtil.getNullIndex(aggregateValue);
-    if (nullIndex != -1) {
-      aggregateValue[nullIndex] = currentValue;
-      Arrays.sort(aggregateValue, comparator);
-      return aggregateValue;
+    Set<T> set = new HashSet<>(currentAggValList);
+    set.add(currentVal);
+    List<T> list = new ArrayList<>(set);
+    list.sort(Comparator.reverseOrder());
+    if (list.size() > tkVal) {
+      list.remove(list.size() - 1);
     }
-
-    aggregateValue[aggregateValue.length - 1] = currentValue;
-    Arrays.sort(aggregateValue, comparator);
-    return aggregateValue;
+    return list;
   }
 
   @SuppressWarnings("unchecked")
   @Override
-  public Merger<String, T[]> getMerger() {
-    return (aggKey, aggOne, aggTwo) -> {
-      final T[] merged = (T[]) Array.newInstance(ttClass, tkVal);
-
-      int idx1 = 0;
-      int idx2 = 0;
-      for (int i = 0; i != tkVal; ++i) {
-        final T v1 = idx1 < aggOne.length ? aggOne[idx1] : null;
-        final T v2 = idx2 < aggTwo.length ? aggTwo[idx2] : null;
-
-        final int result = comparator.compare(v1, v2);
-        if (result < 0) {
-          merged[i] = v1;
-          idx1++;
-        } else if (result == 0) {
-          merged[i] = v1;
-          idx1++;
-          idx2++;
-        } else {
-          merged[i] = v2;
-          idx2++;
-        }
-      }
-
-      return merged;
+  public Merger<String, List<T>> getMerger() {
```

Comment:
Previous flagged as a performance concern. Your impl takes over 6 minutes on my machine to run the test in `IntTopkDistinctKudfTest`. The impl below, based heavily on the code that was already there, takes around 1:20. So yours update is 3.5 times slower.

```
@SuppressWarnings("unchecked")
  @Override
  public Merger<String, List<T>> getMerger() {
    return (aggKey, aggOne, aggTwo) -> {
      // Orig: 1m:21
      final List<T> merged = new ArrayList<>(Math.min(tkVal, aggOne.size() + aggTwo.size()));

      int idx1 = 0;
      int idx2 = 0;
      for (int i = 0; i != tkVal; ++i) {
        final T v1 = idx1 < aggOne.size() ? aggOne.get(idx1) : null;
        final T v2 = idx2 < aggTwo.size() ? aggTwo.get(idx2) : null;

        if (v1 == null && v2 == null) {
          break;
        }

        if (v1 != null && (v2 == null || v1.compareTo(v2) > 0)) {
          merged.add(v1);
          idx1++;
        } else if (v1 == null || v2.compareTo(v1) < 0) {
          merged.add(v2);
          idx2++;
        } else if (v1.compareTo(v2) == 0) {
          merged.add(v1);
          idx1++;
          idx2++;
        }
      }
      return merged;
}
```

Reply:
I updated the code with the above one.